[GeoMechanicsApplication] Fixed several minor code smells

avdg81 opened the pull request:
**📝 Description**
Fixed several issues reported by SonarQube and Clang-tidy.

**🆕 Changelog**
- Use free functions `std::begin` and `std::end` to iterate over generic ranges.
- Replaced a template parameter by class `Geometry<Node>`, since that appeared to be the intended use case.
- ~~Made several member functions `static`.~~ Made several members [free functions](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c4-make-a-function-a-member-only-if-it-needs-direct-access-to-the-representation-of-a-class) in an [unnamed namespace](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#sf22-use-an-unnamed-anonymous-namespace-for-all-internalnon-exported-entities) to comply with the C++ Core Guidelines.
- Marked a few member functions `[[nodiscard]]`.
- Adapted some test code to fix build warnings.
- Moved an `#include` from a header file to the implementation file.

avdg81 committed:
Fixed several minor code smells

- Use free functions `std::begin` and `std::end` to iterate over generic ranges.
- Replaced a template parameter by class `Geometry<Node>`, since that appeared to be the intended use case.
- Made several member functions `static`.
- Marked a few member functions `[[nodiscard]]`.
- Adapted some test code to fix build warnings.
- Moved an `#include` from a header file to the implementation file.

rfaasse commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -19,6 +19,8 @@
 #include "custom_utilities/stress_strain_utilities.h"
 #include "geo_mechanics_application_variables.h"
 
+#include <cmath>
```
I might have missed it, but I don't see any new code here that requires this include

rfaasse replied:
Or was it missing and does e.g. std::tan (line 242) need it?

avdg81 replied:
Previously, it was included in the header file, but it was only needed in the implementation file. That's why I've moved it to here. It is needed for `std::tan`, `std::sin`, and `std::cos`.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/generic_utilities.h:
```diff
@@ -31,7 +31,7 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GenericUtilities
     {
         auto result = VectorType(rVector.size());
-        std::transform(rIndices.begin(), rIndices.end(), result.begin(),
+        std::transform(std::begin(rIndices), std::end(rIndices), result.begin(),
```
Just out of curiosity, what is the advantage of this over calling .begin()? Is it to make the templated function even more generic and also have it work on containers which don't have a begin(), but std::begin can find the iterator?

avdg81 replied:
My understanding is that [`std::begin`](https://en.cppreference.com/w/cpp/iterator/begin) is slightly more generic than assuming that the supplied range has a member function `begin`. For instance, it means that you could also pass it a C-style array, which would not work with the previous implementation. Since this is generic code, I feel that the issue reported by SonarQube is a valid one.

rfaasse reviewed (dismissed):
Thanks for fixing these these code smells! No blockers from my side, just a few questions out of curiosity

avdg81 committed:
Made some `static` members free functions

This is in line with C++ Core Guideline C.4: Make a function a member only if it needs direct access to the representation of a class. The free functions have been put in an unnamed namespace to comply with C++ Core Guideline SF.22: Use an unnamed (anonymous) namespace for all internal/non-exported entities.

rfaasse reviewed (approved):
Looks good to me!

mnabideltares commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -19,19 +19,81 @@
-    result[0]   = 0.5 * (rPrincipalStresses(0) + rPrincipalStresses(2));
-    result[1]   = 0.5 * (rPrincipalStresses(0) - rPrincipalStresses(2));
+    result[0]   = 0.5 * (rPrincipalStresses[0] + rPrincipalStresses[2]);
+    result[1]   = 0.5 * (rPrincipalStresses[0] - rPrincipalStresses[2]);
```
Is there any reason to change () to [] ? Don't we keep by convension [] for std:vector and () for Vector?
Because for Matrix we need to use (,) which will be inconsistent with Vector from the same library (Ublas-boost)

avdg81 replied:
I thought that the convention was to use `operator[]` for indexing operations with a single index (this applies to `std::vector` as well as `Vector`). If we'd use different operators for the same operation (semantically speaking) depending on the underlying type would be rather confusing.

`Matrix` on the other hand doesn't have `operator []`, so we have to use `operator()` there.

mnabideltares reviewed (commented):
@avdg81  Thank you for fixing code smells and clang-tidy issues. 
Removing from static to free functions is more logical here as these functions are specific for this class and they are never used by the other parts of the code.
I have just a minor comment.
